### PR TITLE
refactor: implement mutation backends and migrate handlers

### PR DESCRIPTION
**Pull request by retran**
## Why

PR #236 defined the mutation interfaces — this PR provides the MPR-backed implementations that actually perform the BSON manipulation. The goal is to move all BSON mutation logic out of the executor and behind the backend abstraction, so the executor becomes a thin orchestrator.

## What changed (incremental from PR #236)

**28 files changed, +4,511/-3,523**

**New backend implementations:**
- `mdl/backend/mpr/page_mutator.go` (1,554 lines) — full `PageMutator` implementation: widget tree traversal, property updates, widget add/drop/move, layout changes, pluggable widget operations, all operating on BSON documents
- `mdl/backend/mpr/workflow_mutator.go` (771 lines) — full `WorkflowMutator` implementation: activity insertion/removal, outcome management, path operations, boundary events
- `mdl/backend/mpr/widget_builder.go` (1,007 lines) — widget construction and serialization: `WidgetObjectBuilder`, property type resolution, default widget scaffolding

**Executor simplification:**
- `cmd_alter_page.go`: 1,721 → 256 lines — rewritten as thin orchestrator calling `PageMutator` methods
- `cmd_alter_workflow.go`: 887 → 178 lines — same pattern with `WorkflowMutator`
- `widget_engine.go` refactored to use `WidgetObjectBuilder` interface instead of direct BSON construction
- Deleted from executor: `widget_defaults.go`, `widget_operations.go`, `widget_templates.go` (logic moved to backend)

**Interface integration from #236:**
- Uses typed `ContainerKind`, `InsertPosition`, `PluggablePropertyOp` constants defined in PR #236
- `InsertWidget` signature uses `InsertPosition` typed parameter (case-insensitive comparison in implementation)

**Tests relocated:** ALTER PAGE and ALTER WORKFLOW tests moved from executor to `mdl/backend/mpr/` where the implementation now lives.

**Review-driven improvements (post-PR creation):**
- `NewWidgetRegistryWithOps(extraOps)` — reintroduces extensibility injection point for user-defined widget operations that was lost when `OperationRegistry` was replaced by a fixed `knownOperations` set. `NewWidgetRegistry()` remains the zero-config default.
- Round-trip and wrapper tests for `convert/unconvert` layer in `mdl/backend/mpr/convert_roundtrip_test.go` — 33 tests covering all forward conversion, reverse unconvert, nil/error passthrough, and Slice/Ptr wrapper functions.

## Stack

PR 3/5 in the shared-types refactoring stack. This is the largest PR in the stack.
Merge order: #232 → #235 → #236 → **this** → #238 → #239

### CODE REVIEW

**Comment by ako** [comment]
## Code Review

### Overview

This is the largest PR in the stack (+12,710/-5,703 lines, 28 files) and delivers the actual implementation behind the interfaces defined in PR #236. The executor's `cmd_alter_page.go` shrinks from 1,721 → 256 lines and `cmd_alter_workflow.go` from 887 → 178 lines. Both become thin orchestrators. The BSON mutation logic moves to `page_mutator.go` (1,554 lines), `workflow_mutator.go` (771 lines), and `widget_builder.go` (1,007 lines) in `mdl/backend/mpr/`. This is a well-executed refactoring overall.

*Note: This PR is currently in DRAFT state.*

---

### Positive Aspects

- **Compile-time interface satisfaction checks are present.** Both `page_mutator.go` and `workflow_mutator.go` have `var _ backend.PageMutator = (*mprPageMutator)(nil)` guards — exactly what was flagged as missing in the PR #236 review.
- **BSON tree walking is thorough.** `findBsonWidget` and `findBsonWidgetInSnippet` correctly handle all container types: FormCall, LayoutGrid (rows/columns), TabContainer, DataView (Widgets + FooterWidgets), ControlBar, CustomWidget (pluggable properties). No container type is obviously missing.
- **Widget builder uses templates, not hand-constructed BSON.** `widget_builder.go` loads pre-built templates via `widgets.GetTemplateFullBSON()` and modifies them — which is the safe approach established in the codebase.
- **Thin orchestrators are genuinely thin.** `cmd_alter_page.go` is a clean dispatcher with no BSON logic. `cmd_alter_workflow.go` maps all 10+ operation types through a single switch and delegates every case to a mutator call.
- **`page_mutator_test.go` (725 lines) has solid coverage** — widget tree walking, drop/set operations, scope extraction, BSON array marker preservation.

---

### Issues & Concerns

**1. Potential method visibility issue in `workflow_mutator.go` — verify before merge**

The analysis flagged that the `MprBackend` factory method may be defined with a lowercase name (`openWorkflowForMutation` vs the interface's `OpenWorkflowForMutation`). If real, this would be a compile error. Since this is a DRAFT it may be intentional/WIP — but worth explicitly verifying the method name matches the interface before promoting to ready-for-review.

**2. Workflow mutation operations are undertested**

`workflow_mutator_test.go` (484 lines) covers property setting, activity finding, and drop operations. It does **not** test:
- `InsertAfterActivity`
- `ReplaceActivity`
- `InsertOutcome` / `DropOutcome`
- `InsertPath` / `DropPath`
- `InsertBranch` / `DropBranch`
- `InsertBoundaryEvent`
- `SetActivityProperty` (only one test for DUE_DATE)

These are the most complex mutation operations and the most likely to have edge case bugs (e.g., position handling, sub-flow construction, deduplication). The corresponding `page_mutator_test.go` has better coverage of its operations.

**3. BSON helpers are duplicated across three files**

`bson_helpers.go` (executor-private), `page_mutator.go`, and `workflow_mutator.go` all define the same low-level helpers (`dGet`, `dGetDoc`, `dSet`, `dGetArrayElements`, etc.). This is pragmatic but creates a maintenance risk — a fix to one copy won't propagate. These belong in `mdl/bsonutil` or a shared internal package.

**4. No `MockPageMutator` / `MockWorkflowMutator` types**

`mock_mutation.go` mocks the *backend factory methods* but doesn't provide types that implement the mutator interfaces themselves. The test files construct `mprPageMutator`/`mprWorkflowMutator` directly (testing the real implementation), which is fine for backend tests. But executor-level tests that want to test `cmd_alter_page.go` logic in isolation (without BSON roundtrips) have no mock mutator to inject.

**5. `SetLayout` (96 lines in `page_mutator.go`) warrants close review**

The parameter remapping logic when changing a page's layout is the most complex method in the mutator. It correctly handles old/new layout qualified names and argument list remapping, but the complexity makes it a candidate for dedicated tests — none exist currently.

---

### Checklist Against Project Standards

| Item | Status |
|---|---|
| Interface satisfaction compile-time checks | ✅ Present in both mutator files |
| Thin orchestrators | ✅ Both executor files genuinely delegate all logic |
| BSON container type coverage | ✅ All known container types handled |
| Widget builder uses templates | ✅ No hand-constructed BSON |
| `page_mutator_test.go` coverage | ✅ Tree walking, drops, properties, scope |
| `workflow_mutator_test.go` coverage | ⚠️ Missing insert/replace/outcome/path/branch tests |
| BSON helper deduplication | ⚠️ Three copies across executor and backend |
| Mock mutator types | ⚠️ Factory mocked; mutator interface not mockable |

---

### Verdict

**Strong work overall — the refactoring goal is achieved.** The executor is genuinely decoupled from BSON, the implementations are thorough, and the page-side tests are solid. The main gaps before this moves out of DRAFT are: (1) verify the `OpenWorkflowForMutation` method name; (2) add tests for the write-path workflow operations (insert/replace/outcomes/paths/branches) — these are the riskiest operations and currently have zero test coverage.

File: mdl/executor/cmd_pages_builder.go [hunk 1]
```diff
@@ -305,7 +316,7 @@ func (pb *pageBuilder) resolveFolder(folderPath string) (model.ID, error) {
 			currentContainerID = newFolderID
 
 			// Add to cache
-			pb.foldersCache = append(pb.foldersCache, &mpr.FolderInfo{
+			pb.foldersCache = append(pb.foldersCache, &types.FolderInfo{
 				ID:          newFolderID,
 				ContainerID: currentContainerID,
 				Name:        part,
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_pages_builder.go:323]: Already fixed — line 316 captures `parentContainerID := currentContainerID` before updating `currentContainerID` to `newFolderID`. The `ContainerID` on line 322 correctly uses the stored parent.

File: mdl/executor/cmd_pages_builder_v3_pluggable.go [hunk 2]
```diff
@@ -89,7 +90,15 @@ func (pb *pageBuilder) buildWidgetV3ToBSON(w *ast.WidgetV3) (bson.D, error) {
...
-	return mpr.SerializeWidget(widget), nil
+	raw := pb.widgetBackend.SerializeWidgetToOpaque(widget)
+	if raw == nil {
+		return nil, nil
+	}
+	bsonD, ok := raw.(bson.D)
+	if !ok {
+		return nil, mdlerrors.NewValidationf("SerializeWidgetToOpaque returned unexpected type %T", raw)
+	}
+	return bsonD, nil
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_pages_builder_v3_pluggable.go:101]: By design. The interface returns `any` (opaque contract) so `mdl/backend` stays BSON-free. `buildWidgetV3ToBSON` is the intentional bridge that casts to `bson.D` with proper error handling (lines 97-99). Changing the interface to return `bson.D` would couple the abstraction to a storage format.

File: mdl/types/asyncapi.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+	doc := &AsyncAPIDocument{
+		Version:     raw.AsyncAPI,
+		Title:       raw.Info.Title,
+		DocVersion:  raw.Info.Version,
+		Description: raw.Info.Description,
+	}
+
+	// Resolve messages from components
+	for name, msg := range raw.Components.Messages {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/asyncapi.go]: Already sorted — lines 66-70 collect keys into a slice and `sort.Strings(messageNames)` before iteration.

File: mdl/types/asyncapi.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		} else if msg.Payload.Properties != nil {
+			// Inline schema
+			resolved.Properties = resolveAsyncSchemaProperties(msg.Payload)
+		}
+
+		doc.Messages = append(doc.Messages, resolved)
+	}
+
+	// Resolve channels
+	for channelName, channel := range raw.Channels {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/asyncapi.go]: Already sorted — lines 99-103 collect keys and `sort.Strings(channelNames)` before iteration.

File: mdl/types/asyncapi.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+func asyncRefName(ref string) string {
+	if idx := strings.LastIndex(ref, "/"); idx >= 0 {
+		return ref[idx+1:]
+	}
+	return ref
+}
+
+func resolveAsyncSchemaProperties(schema yamlAsyncSchema) []*AsyncAPIProperty {
+	var props []*AsyncAPIProperty
+	for name, prop := range schema.Properties {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/asyncapi.go]: Already sorted — `resolveAsyncSchemaProperties` (lines 154-158) collects keys and `sort.Strings(names)` before iteration.

File: mdl/executor/widget_registry.go [hunk 6]
```diff
@@ -18,23 +18,27 @@ import (
...
+	"texttemplate":     true,
+	"action":           true,
+	"attributeObjects": true,
 }
 
-// NewWidgetRegistryWithOps creates a registry pre-loaded with embedded definitions,
-// validating operations against the provided OperationRegistry.
-func NewWidgetRegistryWithOps(opReg *OperationRegistry) (*WidgetRegistry, error) {
+// NewWidgetRegistry creates a registry pre-loaded with embedded definitions.
+func NewWidgetRegistry() (*WidgetRegistry, error) {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_registry.go]: Fixed in commit 57261b3 — reintroduced `NewWidgetRegistryWithOps(extraOps)`. User-defined widget definitions can now declare custom operations via the `extraOps` parameter. `NewWidgetRegistry()` remains the zero-config default. Test: `TestNewWidgetRegistryWithOps_ExtendsKnownOperations`.

File: mdl/types/asyncapi.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+// ParseAsyncAPI parses an AsyncAPI YAML string into an AsyncAPIDocument.
+func ParseAsyncAPI(yamlStr string) (*AsyncAPIDocument, error) {
+	if yamlStr == "" {
+		return nil, fmt.Errorf("empty AsyncAPI document")
+	}
+
+	var raw yamlAsyncAPI
+	if err := yaml.Unmarshal([]byte(yamlStr), &raw); err != nil {
+		return nil, fmt.Errorf("failed to parse AsyncAPI YAML: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/asyncapi.go:56]: Tests already exist in `mdl/types/asyncapi_test.go`: `TestParseAsyncAPI_Basic`, `TestParseAsyncAPI_Empty`, `TestParseAsyncAPI_InvalidYAML`, `TestParseAsyncAPI_MultipleChannels_Sorted`, `TestParseAsyncAPI_InlinePayload`, `TestFindMessage` — covering empty input, YAML failures, \$ref resolution, channel/message extraction, sorted ordering, and inline payloads.

**Comment by retran** [comment]
Thanks for the thorough review! Addressing each point:

**#1 (Method visibility):** Verified — no issue. `OpenWorkflowForMutation` (exported) is defined on `MprBackend` in `backend.go:741` and delegates to `openWorkflowForMutation` (unexported) in `workflow_mutator.go:36`. Both exist, compiles and tests pass.

**#2 (Workflow mutation tests):** Fixed in `3b67e04`. Added 48+ tests to `workflow_mutator_test.go` covering all previously untested operations: InsertAfterActivity, ReplaceActivity, InsertOutcome/DropOutcome, InsertPath/DropPath, InsertBranch/DropBranch (true/false/default/enum conditions), InsertBoundaryEvent (InterruptingTimer, NonInterruptingTimer, Timer with activities), and expanded SetActivityProperty (PAGE, DESCRIPTION, TARGETING_MICROFLOW, TARGETING_XPATH) plus SetPropertyWithEntity (OVERVIEW_PAGE set/clear, PARAMETER new/update/clear).

Also discovered and documented a `dSet` bug: it only updates existing BSON keys, never creates new ones. `SetActivityProperty` for PAGE/TARGETING_MICROFLOW/TARGETING_XPATH silently fails when the target key does not pre-exist. See `TestWorkflowMutator_SetActivityProperty_Page_MissingKey`. Tracked for a follow-up fix.

**#3 (BSON helpers dedup):** Already addressed in PR #238 — `mdl/executor/bson_helpers.go` is removed in `feature/decouple-executor`. The remaining copy in `page_mutator.go` is shared within `mprbackend` package by `workflow_mutator.go`. No action needed in #237.

**#4 (MockPageMutator/MockWorkflowMutator):** Deferred. These ~30-method FuncField mock types would enable executor-level isolation tests for `cmd_alter_page.go`/`cmd_alter_workflow.go`, but the current PR tests exercise the real mpr implementation directly, which gives stronger guarantees. Tracked as a future item.

**#5 (SetLayout tests):** Fixed in `3b67e04`. Added 6 dedicated tests to `page_mutator_test.go`: basic layout change with parameter remapping, explicit param mappings, same layout no-op, snippet rejection, missing FormCall, and empty Form field.

File: mdl/executor/cmd_alter_workflow.go [hunk 8]
```diff
@@ -61,84 +51,117 @@ func execAlterWorkflow(ctx *ExecContext, s *ast.AlterWorkflowStmt) error {
...
 		case *ast.SetWorkflowPropertyOp:
-			if err := applySetWorkflowProperty(&rawData, o); err != nil {
-				return mdlerrors.NewBackend("SET "+o.Property, err)
+			switch o.Property {
+			case "OVERVIEW_PAGE", "PARAMETER":
+				qn := o.Entity.Module + "." + o.Entity.Name
+				if qn == "." {
+					qn = ""
+				}
+				if err := mutator.SetPropertyWithEntity(o.Property, o.Value, qn); err != nil {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_alter_workflow.go:80]: Good catch — this is a real bug. For OVERVIEW_PAGE, the parser sets `o.Entity` (the page qualified name) but never sets `o.Value`, so `o.Value` is always empty. This caused every OVERVIEW_PAGE SET to clear AdminPage instead of setting it. Fixed in 32a4230 — OVERVIEW_PAGE now passes `qn` (from Entity) as both value and entity args. PARAMETER handling unchanged (uses `o.Value` as variable name correctly).

File: mdl/executor/widget_registry.go [hunk 9]
```diff
@@ -18,23 +18,53 @@ import (
...
-// validating operations against the provided OperationRegistry.
-func NewWidgetRegistryWithOps(opReg *OperationRegistry) (*WidgetRegistry, error) {
+// extending the default known operations with extraOps for validation.
+// This allows user-defined widgets to declare custom operations that would otherwise
+// fail validation. Pass nil for the default set.
+func NewWidgetRegistryWithOps(extraOps map[string]bool) (*WidgetRegistry, error) {
+	knownOperations = copyOps(defaultKnownOperations)
+	for op := range extraOps {
+		knownOperations[op] = true
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_registry.go:63]: Fixed in 32a4230 — moved `knownOperations` from package-level mutable var to a `knownOperations` field on `WidgetRegistry`. Converted `validateDefinitionOperations` and `validateMappings` from free functions to `WidgetRegistry` methods. No more global state; each registry instance carries its own operations set.

File: mdl/executor/cmd_pages_builder_input_datagrid.go [hunk 10]
```diff
@@ -85,7 +84,7 @@ func (pb *pageBuilder) buildDataGrid2Property(entry pages.PropertyTypeIDEntry, d
 	// Build the datasource BSON if provided
 	var datasourceBSON any
 	if datasource != nil {
-		datasourceBSON = mpr.SerializeCustomWidgetDataSource(datasource)
+		datasourceBSON = pb.widgetBackend.SerializeDataSourceToOpaque(datasource)
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_pages_builder_input_datagrid.go:88]: Valid point. The opaque contract means the caller trusts the backend to return BSON-compatible values. Adding a type check here would be inconsistent — `SerializeWidgetToOpaque` at `cmd_pages_builder_v3_pluggable.go:97` already does the type assertion with error handling for the widget path, but this datasource path embeds the value directly into a `bson.D` element (which accepts `any`). The BSON marshal will surface type errors at serialization time. Adding early validation would require changing `buildDataGrid2Property` to return `(bson.D, error)` and propagating error returns up the caller chain — tracking as a follow-up improvement.

File: mdl/executor/widget_engine.go [hunk 11]
```diff
@@ -500,65 +478,58 @@ func (e *PluggableWidgetEngine) resolveMapping(mapping PropertyMapping, w *ast.W
...
+		ctx := &BuildContext{}
+		if err := e.applyOperation(builder, slot.Operation, slot.PropertyKey, ctx); err != nil {
+			return err
 		}
-
-		ctx := &BuildContext{ChildWidgets: childBSONs}
-		*updatedObject = op(*updatedObject, propertyTypeIDs, slot.PropertyKey, ctx)
+		// SetChildWidgets directly — applyOperation skips "widgets" since ctx doesn't carry children
+		builder.SetChildWidgets(slot.PropertyKey, children)
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_engine.go:540]: Fixed in da2799b — added validation that `slot.Operation` must be `"widgets"`. Non-widgets operations now return a clear error instead of silently running the operation and then overwriting with `SetChildWidgets`. All existing definitions use `"widgets"` for child slots, so this is a defensive guard for future definitions.

File: mdl/backend/mpr/workflow_mutator.go [hunk 12]
```diff
@@ -0,0 +1,771 @@
...
+				{Key: "Name", Value: "WorkflowContext"},
+			}
+			for i, elem := range m.rawData {
+				if elem.Key == "Parameter" {
+					m.rawData[i].Value = newParam
+					return nil
+				}
+			}
+			m.rawData = append(m.rawData, bson.E{Key: "Parameter", Value: newParam})
+		}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/workflow_mutator.go:142]: By design — in Mendix, workflows have exactly one parameter always named "WorkflowContext" (see grammar: `PARAMETER $WorkflowContext: MyModule.Request`). The `value` argument carries the parameter name for the SET/CLEAR signal, but the stored name is always "WorkflowContext" because that is the Mendix domain constraint. Using a different name would produce an invalid model.

File: mdl/executor/bson_helpers.go [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+	}
+	dSet(doc, key, result)
+}
+
+// extractBinaryIDFromDoc extracts a binary ID string from a bson.D field.
+func extractBinaryIDFromDoc(val any) string {
+	if bin, ok := val.(primitive.Binary); ok {
+		return types.BlobToUUID(bin.Data)
+	}
+	return ""
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/bson_helpers.go]: Good catch — fixed in da2799b. `extractBinaryIDFromDoc` now handles both `primitive.Binary` and `[]byte` via type switch, consistent with `matchesTypePointer` which already supported both.

**Comment by retran** [comment]
Thanks for the thorough review. Addressing the two findings:

**1. `ContainerKind` string comparison (line 257):** Not a bug. `ContainerKind` is `type ContainerKind string` (defined in `mdl/backend/mutation.go:12`), and `ContainerSnippet = "snippet"`. Go allows comparing a named string type with a string literal — this compiles and works correctly. That said, using the constant `backend.ContainerSnippet` would be more explicit; will update.

**2. `SetLayout` oldLayoutQN detection logic:** Not a bug. The detection works in two stages: (a) the `Form` field stores the full qualified name (e.g. `"MyModule.MyLayout"`), so `oldLayoutQN` gets the full QN; (b) the `Arguments` fallback (`&& oldLayoutQN == ""`) only runs if the Form field was empty, extracting a module name from parameter values as a last resort. The `newLayout` parameter is also a full QN (confirmed by tests: `SetLayout("MyModule.NewLayout", nil)`), so the comparison `oldLayoutQN == newLayout` correctly compares full qualified names. The parameter remapping strips the old QN prefix and prepends the new one. All 6 SetLayout tests pass and verify this behavior.

File: mdl/executor/widget_engine.go [hunk 14]
```diff
@@ -149,8 +150,7 @@ func (e *PluggableWidgetEngine) Build(def *WidgetDefinition, w *ast.WidgetV3) (*
 					if err != nil {
 						return nil, mdlerrors.NewBackend("auto datasource for "+propKey, err)
 					}
-					ctx := &BuildContext{DataSource: dataSource, EntityName: entityName}
-					updatedObject = opDatasource(updatedObject, propertyTypeIDs, propKey, ctx)
+					builder.SetDataSource(propKey, dataSource)
 					if entityName != "" {
 						e.pageBuilder.entityContext = entityName
 					}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_engine.go]: Valid concern. The ordering is: (1) applyChildSlots builds children, (2) auto-datasource sets entityContext. If a widget's .def.json omits the DataSource mapping but the AST has a DataSource, children built in step 1 won't have entityContext set yet.

In practice this is rare — widgets with DataSource almost always have it in their .def.json PropertyMappings, so entityContext is set in step 3 before step 4. But the edge case exists. Tracked for fix — will move auto-datasource before applyChildSlots.

File: mdl/executor/bson_helpers.go [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+	default:
+		return setPluggableWidgetProperty(widget, propName, value)
+	}
+}
+
+func setWidgetCaption(widget bson.D, value interface{}) error {
+	caption := dGetDoc(widget, "Caption")
+	if caption == nil {
+		setTranslatableText(widget, "Caption", value)
+		return nil
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/bson_helpers.go]: Valid point. When the widget has no Caption sub-document, `setTranslatableText` falls back to `dSet(widget, "Caption", strVal)` which injects a plain string field — incorrect for widgets expecting a structured translatable object. Will fix to match `setWidgetContent` behavior: return a validation error when Caption structure is missing.

File: mdl/executor/widget_engine.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_engine.go]: Fixed in 8d8017f — moved auto-datasource resolution before `applyChildSlots` so that `entityContext` is available for child widgets that depend on the parent's data source.

File: mdl/executor/bson_helpers.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/bson_helpers.go]: Fixed in 8d8017f — `setWidgetCaption` now returns a validation error when the widget has no Caption property, matching `setWidgetContent` behavior. No more silent plain-string injection.

File: mdl/executor/widget_engine.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_engine.go]: Fixed in 10f460d — auto-datasource now runs before applyChildSlots, so entityContext is available for child widgets.

File: mdl/executor/bson_helpers.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/bson_helpers.go]: Fixed in 10f460d — setWidgetCaption now returns `mdlerrors.NewValidation("widget has no Caption property")` when Caption is nil, matching setWidgetContent behavior.

File: mdl/backend/mpr/workflow_mutator.go [hunk 16]
```diff
@@ -0,0 +1,771 @@
...
+	if err != nil {
+		return err
+	}
+
+	events := dGetArrayElements(dGet(actDoc, "BoundaryEvents"))
+	if len(events) == 0 {
+		return fmt.Errorf("activity %q has no boundary events", activityRef)
+	}
+
+	// Drop the first boundary event silently.
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/workflow_mutator.go:536]: By design — MDL grammar only supports one boundary event per activity (`BoundaryEvent <EventType> [Delay <expr>] [Recurrence <expr>]` — singular, not a list). The BSON array structure is a storage artifact; the model always has 0 or 1 boundary event. Dropping the first element is correct behavior. The comment says 'silently' to distinguish from the 'no events' error case above.

File: mdl/backend/mpr/widget_builder.go [hunk 17]
```diff
@@ -0,0 +1,1007 @@
...
+			} else {
+				result[i] = item
+			}
+		} else {
+			result[i] = item
+		}
+	}
+	if !matched {
+		log.Printf("WARNING: updatePropertyInArray: no match for TypePointer %s in %d properties", propertyTypeID, len(arr)-1)
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/widget_builder.go:367]: This is pre-existing code moved from the executor package — the log.Printf is a diagnostic warning for widget template development, not user-facing output. Changing it to return an error would break widgets that legitimately don't have all TypePointers present (optional properties). The warning helps developers identify mapping mismatches during development. No change needed.

**Comment by retran** [comment]
Thanks for the review. Addressing each point:

**Moderate — `serializeWidgets` skips nil widgets:** By design. `mpr.SerializeWidget` returns nil for widget types that have no BSON representation (e.g., placeholder widgets, layout-only containers). These are structural AST nodes that don't produce output in the serialized model. Skipping them is correct — returning an error would break valid page structures. The caller (`InsertWidget`, `SetChildWidgets`) expects that the resulting slice may be shorter than the input.

**Minor — PR size:** Acknowledged. This is the largest PR in a 5-PR stack (#235→#236→#237→#238→#239). Each PR is scoped to a single concern, and this one is the implementation that backs the interfaces defined in #236. The size is a consequence of moving ~50 handler functions from executor to backend — splitting further would create partial migrations that don't compile independently.

**Minor — `SetPluggableProperty` switch statement:** Agreed it could grow. The operations are currently finite and stable (defined by the `.def.json` spec). If new operation types are added, we can refactor to a registry pattern (similar to `WidgetRegistry.knownOperations`). Acceptable for now.

File: mdl/backend/mpr/widget_builder.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/widget_builder.go:367]: Pre-existing diagnostic log moved from the executor package. Changing to return error would break the widget building contract where missing optional properties are expected. The log is gated behind WARNING level — it's a development-time signal for widget template authors, not user-facing output.

File: mdl/backend/mock/mock_microflow.go [hunk 18]
```diff
@@ -49,6 +49,13 @@ func (m *MockBackend) MoveMicroflow(mf *microflows.Microflow) error {
 	return nil
 }
 
+func (m *MockBackend) ParseMicroflowFromRaw(raw map[string]any, unitID, containerID model.ID) *microflows.Microflow {
+	if m.ParseMicroflowFromRawFunc != nil {
+		return m.ParseMicroflowFromRawFunc(raw, unitID, containerID)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mock/mock_microflow.go]: Fixed in a2394be — mock now panics with a clear message when ParseMicroflowFromRawFunc is not set, making test failures immediately diagnosable.

File: mdl/backend/mutation.go [hunk 19]
```diff
@@ -63,14 +80,19 @@ type PageMutator interface {
...
-	// DropWidget removes widgets by name from the tree.
-	DropWidget(widgetRefs []string) error
+	// DropWidget removes widgets by ref from the tree.
+	DropWidget(refs []WidgetRef) error
 
-	// ReplaceWidget replaces the target widget with the given widgets.
-	ReplaceWidget(targetWidget string, widgets []pages.Widget) error
+	// ReplaceWidget replaces the target widget or column with the given widgets.
+	// columnRef is "" for widget targeting.
+	ReplaceWidget(widgetRef string, columnRef string, widgets []pages.Widget) error
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mutation.go:92]: Valid API design concern. The current split exists because InsertWidget/ReplaceWidget need to target both widget-level and column-level positions (columnRef disambiguates), while DropWidget only needs widget identification. Unifying to WidgetRef would require WidgetRef to carry an optional column field — tracked for a follow-up API cleanup.

File: mdl/backend/mpr/widget_builder.go [hunk 20]
```diff
@@ -0,0 +1,1007 @@
...
+
+	ob.object = updateWidgetPropertyValue(ob.object, ob.propertyTypeIDs, propertyKey, func(val bson.D) bson.D {
+		// The val here is the WidgetValue — but we need the WidgetProperty level.
+		// CloneGallerySelectionProperty works on the Properties array directly.
+		return val
+	})
+
+	// Actually need to work at the Properties array level: find the property,
+	// clone it with new IDs and updated Selection, then append.
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/widget_builder.go]: Fixed in a2394be — removed the no-op `updateWidgetPropertyValue` call. CloneGallerySelectionProperty now works directly at the Properties array level without the redundant walk.

File: mdl/backend/mpr/workflow_mutator.go [hunk 21]
```diff
@@ -0,0 +1,771 @@
...
+		}
+	}
+}
+
+// deduplicateNewActivityName ensures a new activity name doesn't conflict.
+func deduplicateNewActivityName(act workflows.WorkflowActivity, existingNames map[string]bool) {
+	name := act.GetName()
+	if name == "" || !existingNames[name] {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/workflow_mutator.go]: Good catch — real bug. Fixed in a2394be. When the name is unique, it's now recorded in `existingNames` before returning, so batch inserts with duplicate names are properly deduplicated.

File: mdl/executor/widget_engine.go [hunk 22]
```diff
@@ -160,43 +157,43 @@ func (e *PluggableWidgetEngine) Build(def *WidgetDefinition, w *ast.WidgetV3) (*
...
 	for _, s := range slots {
 		handledSlotKeys[s.PropertyKey] = true
 	}
-	// Collect Widgets-type property keys
 	var widgetsPropKeys []string
 	for propKey, entry := range propertyTypeIDs {
 		if entry.ValueType == "Widgets" && !handledSlotKeys[propKey] {
 			widgetsPropKeys = append(widgetsPropKeys, propKey)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/widget_engine.go:176]: Fixed in a2394be — added `sort.Strings(widgetsPropKeys)` after building the slice from the map, ensuring deterministic default-slot assignment.
